src/include/storage/index/index.h

```diff
@@ -50,46 +63,84 @@ class Index {
   virtual ~Index() = default;
 
   /**
-   * Inserts a new key-value pair into the index.
+   * Inserts a new key-value pair into the index, used for non-unique key indexes.
+   * @param txn txn context for the calling txn, used to register abort actions
    * @param tuple key
    * @param location value
    * @return false if the value already exists, true otherwise
    */
-  virtual bool Insert(const ProjectedRow &tuple, TupleSlot location) = 0;
+  virtual bool Insert(transaction::TransactionContext *txn, const ProjectedRow &tuple, TupleSlot location) = 0;
 
   /**
-   * Removes a key-value pair from the index.
+   * Inserts a key-value pair only if any matching keys have TupleSlots that don't conflict
+   * @param txn txn context for the calling txn, used for visibility and write-write, and to register abort actions
    * @param tuple key
    * @param location value
-   * @return false if the key-value pair did not exist, true if the deletion succeeds
+   * @return true if the value was inserted, false otherwise
+   *         (either because value exists, or predicate returns true for one of the existing values)
    */
-  virtual bool Delete(const ProjectedRow &tuple, TupleSlot location) = 0;
+  virtual bool InsertUnique(transaction::TransactionContext *txn, const ProjectedRow &tuple, TupleSlot location) = 0;
 
   /**
-   * Inserts a key-value pair only if the predicate fails on all existing values.
+   * Doesn't immediately call delete on the index. Registers a commit action in the txn that will eventually register a
+   * deferred action for the GC to safely call delete on the index when no more transactions need to access the key.
```

Comment:
Should this be a warning because of the implications on when it would be safe to deallocate the index?  Specifically, the deferral of the delete is not atomic with the commit itself and therefore may trigger later than when you expect.